server/src/main/java/org/elasticsearch/indices/IndicesWriteLoadStatsService.java

```diff
@@ -0,0 +1,201 @@
+        this.indicesWriteLoadStore = indicesWriteLoadStore;
+        this.threadPool = threadPool;
+        this.samplingFrequency = SAMPLING_FREQUENCY_SETTING.get(settings);
+        this.storeFrequency = STORE_FREQUENCY_SETTING.get(settings);
+        this.enabled = ENABLED_SETTING.get(settings);
+
+        clusterSettings.addSettingsUpdateConsumer(SAMPLING_FREQUENCY_SETTING, this::setSamplingFrequency);
+        clusterSettings.addSettingsUpdateConsumer(STORE_FREQUENCY_SETTING, this::setStoreFrequency);
+        clusterSettings.addSettingsUpdateConsumer(ENABLED_SETTING, this::setEnabled);
+    }
+
+    @Override
+    protected void doStart() {
+        if (started.compareAndSet(false, true)) {
+            maybeScheduleTasks();
+        }
+    }
+
+    @Override
+    protected void doStop() {
+        if (started.compareAndSet(true, false)) {
+            enabled = false;
+            maybeCancelTasks();
+        }
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+        indicesWriteLoadStore.close();
+    }
+
+    private void setSamplingFrequency(TimeValue samplingFrequency) {
+        this.samplingFrequency = samplingFrequency;
+    }
+
+    private void setStoreFrequency(TimeValue storeFrequency) {
+        this.storeFrequency = storeFrequency;
+    }
+
+    private void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        if (enabled) {
+            maybeScheduleTasks();
+        } else {
+            maybeCancelTasks();
+        }
+    }
+
+    private void collectWriteLoadSamples() {
+        if (enabled == false) {
+            return;
+        }
+
+        try {
+            indexShardWriteLoadStatsCollector.collectWriteLoadStats();
+        } catch (Exception e) {
+            logger.warn("Unable to collect write load stats", e);
+        }
+
+        maybeScheduleSampling();
+    }
+
+    private void storeWriteLoadDistributions() {
+        if (enabled == false) {
+            return;
+        }
+
+        try {
+            final var writeLoadDistributions = indexShardWriteLoadStatsCollector.getWriteLoadDistributionAndReset();
+            indicesWriteLoadStore.putAsync(writeLoadDistributions);
+        } catch (Exception e) {
+            logger.warn("Unable to store shard write load distributions", e);
+        }
+
+        maybeScheduleStore();
+    }
+
+    private void maybeScheduleTasks() {
+        maybeScheduleSampling();
+        maybeScheduleStore();
+    }
+
+    private void maybeCancelTasks() {
+        if (scheduledSampling != null) {
+            scheduledSampling.cancel();
+        }
+
+        if (scheduledStore != null) {
+            scheduledStore.cancel();
+        }
+    }
+
+    private void maybeScheduleSampling() {
+        if (enabled == false) {
+            return;
+        }
+
+        scheduledSampling = threadPool.schedule(this::collectWriteLoadSamples, samplingFrequency, ThreadPool.Names.GENERIC);
```

Comment:
We can discuss this later too, no need to fix in this PR. But I wonder if we should just have a single special thread for this rather than using the generic pool?